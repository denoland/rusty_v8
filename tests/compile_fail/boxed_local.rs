// Copyright 2019-2020 the Deno authors. All rights reserved. MIT license.
use std::pin::pin;

pub fn main() {
  let mut isolate = v8::Isolate::new(mock());
  let scope1 = pin!(v8::HandleScope::new(&mut isolate));
  let mut scope1 = scope1.init();

  let _boxed_local = {
    let scope = pin!(v8::HandleScope::new(&mut scope1));
    let mut scope = scope.init();
    let scope3 = pin!(v8::HandleScope::new(&mut scope));
    let mut scope3 = scope3.init();
    Box::new(v8::Integer::new(&mut scope3, 123))
  };
}

fn mock<T>() -> T {
  unimplemented!()
}
