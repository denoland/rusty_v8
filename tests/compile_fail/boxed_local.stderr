error[E0597]: `scope2` does not live long enough
  --> tests/compile_fail/boxed_local.rs:12:44
   |
9  |   let _boxed_local = {
   |       ------------ borrow later stored here
10 |     let scope2 = pin!(v8::HandleScope::new(&mut scope1));
11 |     let mut scope2 = scope2.init();
   |         ---------- binding `scope2` declared here
12 |     let scope3 = pin!(v8::HandleScope::new(&mut scope2));
   |                                            ^^^^^^^^^^^ borrowed value does not live long enough
...
15 |   };
   |   - `scope2` dropped here while still borrowed

error[E0716]: temporary value dropped while borrowed
  --> tests/compile_fail/boxed_local.rs:12:18
   |
9  |   let _boxed_local = {
   |       ------------ borrow later stored here
...
12 |     let scope3 = pin!(v8::HandleScope::new(&mut scope2));
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ creates a temporary value which is freed while still in use
...
15 |   };
   |   - temporary value is freed at the end of this statement
   |
   = note: consider using a `let` binding to create a longer lived value
   = note: this error originates in the macro `pin` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0716]: temporary value dropped while borrowed
  --> tests/compile_fail/boxed_local.rs:10:18
   |
9  |   let _boxed_local = {
   |       ------------ borrow later stored here
10 |     let scope2 = pin!(v8::HandleScope::new(&mut scope1));
   |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ creates a temporary value which is freed while still in use
...
15 |   };
   |   - temporary value is freed at the end of this statement
   |
   = note: consider using a `let` binding to create a longer lived value
   = note: this error originates in the macro `pin` (in Nightly builds, run with -Z macro-backtrace for more info)
